tools/build_config/build_config.bzl

```diff
@@ -1,48 +1,142 @@
 load("@io_bazel_rules_kotlin//kotlin:jvm.bzl", "kt_jvm_library")
 
-_STRING_TYPE = "String"
-_BOOLEAN_TYPE = "Boolean"
-_INT_TYPE = "Int"
-_LONG_TYPE = "Long"
+def _flatten_key_value_pair(keys = [], values = []):
+    result = []
+    for index, key in enumerate(keys):
+        value = values[index]
+        result.append(
+            "{key}={value}".format(
+                key = key,
+                value = value,
+            )
+        )
 
-def _raw_string(original):
-    """
-    Wraps the given original string to
-        * Kotlin's raw string literal
-        * Escape $ in Kotlin multi line string
+    return result
+
+def _build_config_generator_impl(ctx):
+    package_name = ctx.attr.package_name
+    string_keys = ctx.attr.string_keys
+    string_values = ctx.attr.string_values
+    boolean_keys = ctx.attr.boolean_keys
+    boolean_values = ctx.attr.boolean_values
+    int_keys = ctx.attr.int_keys
+    int_values = ctx.attr.int_values
+    long_keys = ctx.attr.long_keys
+    long_values = ctx.attr.long_values
+
+    strings = _flatten_key_value_pair(string_keys, string_values)
+    booleans = _flatten_key_value_pair(boolean_keys, boolean_values)
+    ints = _flatten_key_value_pair(int_keys, int_values)
+    longs = _flatten_key_value_pair(long_keys, long_values)
+
+    args = ctx.actions.args()
+    args.set_param_file_format("multiline")
+    args.use_param_file("--flagfile=%s", use_always = True)
+
+    args.add("--package", package_name)
+    args.add_joined(
+        "--strings",
+        strings,
+        join_with = ","
+    )
+    args.add_joined(
+        "--booleans",
+        booleans,
+        join_with = ","
+    )
+    args.add_joined(
+        "--ints",
+        ints,
+        join_with = ","
+    )
+    args.add_joined(
+        "--longs",
+        longs,
+        join_with = ","
+    )
 
-    See: https://kotlinlang.org/docs/reference/basic-types.html#string-templates
-    """
-    return '"""' + original.replace("$", "${'$'}") + '"""'
+    output_directory = ctx.actions.declare_directory(ctx.label.name)
```

Comment:
This directory is required so that the path can be passed over to the tool which uses javapoet to generate the java file.

It also seems like javapoet is unable to write to a declared file and only able to write to a directory where it will produce our desired file in a directory structure that follows the java convention (i.e. package name structure)

Reply:
Updated to using a single `declare_file`